lib/srv/forward/sshserver.go

```diff
@@ -861,6 +863,184 @@ func (s *Server) handleAgentForward(ch ssh.Channel, req *ssh.Request, ctx *srv.S
 	return nil
 }
 
+// handleX11ChannelRequest accepts an X11 channel and forwards it back to the client.
+// Servers which support X11 forwarding request a separate channel for serving each
+// inbound connection on the X11 socket of the remote session.
+func (s *Server) handleX11ChannelRequest(ctx context.Context, xreq ssh.NewChannel) error {
+	// accept inbound X11 channel from server
+	sch, sin, err := xreq.Accept()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// setup outbound X11 channel to client
+	cch, cin, err := s.sconn.OpenChannel(sshutils.X11ChannelRequest, xreq.ExtraData())
+	if err != nil {
+		sch.Close()
+		return trace.Wrap(err)
+	}
+
+	// setup child context which will be canceled when io forwarding completes.
+	ctx, cancel := context.WithCancel(ctx)
+
+	// setup wait group for io forwarding goroutines
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// forward data from server to client
+		io.Copy(cch, sch)
+		// inform client that no more data is coming
+		cch.CloseWrite()
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		// forward data from client to server
+		io.Copy(sch, cch)
+		// inform server that no more data is coming
+		sch.CloseWrite()
+	}()
```

Comment:
Nothing calls `cch.Close()` and `sch.Close()`, are these `CloseWrite` calls sufficient to clean up the connection?

Reply:
Good catch, `CloseWrite` is not sufficient.  Accidentally dropped the `Close` call when refactoring.